File: python/open_guardrail/guards/api_response_time.py
"""Detect unrealistic response time claims and SLA violations."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_UNREALISTIC = [
    re.compile(
        r"\binstant(aneous)?\s+(response|processing|result)",
        re.I,
    ),
    re.compile(
        r"\breal[\s-]?time\b.*"
        r"\b(analys|process|comput|translat)",
        re.I,
    ),
    re.compile(
        r"\b(zero|0)\s*(ms|millisecond|latency)\b",
        re.I,
    ),
    re.compile(r"\bno\s+(delay|latency|wait)\b", re.I),
    re.compile(
        r"\bimmediate(ly)?\s+"
        r"(process|analyz|translat|comput)",
        re.I,
    ),
]

_SLA_VIOLATION = [
    re.compile(
        r"\bguarantee[ds]?\s+(100%\s+)?"
        r"(uptime|availability)",
        re.I,
    ),
    re.compile(
        r"\b(99\.999+|100)%\s+(uptime|availability)",
        re.I,
    ),
    re.compile(
        r"\bnever\s+(goes?\s+down|fail|timeout)",
        re.I,
    ),
    re.compile(
        r"\bzero\s+(downtime|failures)", re.I
    ),
]


class _ApiResponseTime:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "api-response-time"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for p in _UNREALISTIC:
            if p.search(text):
                issues.append("unrealistic_claim")
        for p in _SLA_VIOLATION:
            if p.search(text):
                issues.append("sla_violation")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="api-response-time",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(len(issues) / 3, 1.0)
                if triggered
                else 0.0
            ),
            message=(
                "Unrealistic response time claim"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def api_response_time(
    *, action: str = "block"
) -> _ApiResponseTime:
    return _ApiResponseTime(action=action)

File: packages/guards/src/api-response-time.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ApiResponseTimeOptions {
  action: 'block' | 'warn';
}

const UNREALISTIC_PATTERNS = [
  /\binstant(aneous)?\s+(response|processing|result)/i,
  /\breal[\s-]?time\b.*\b(analys|process|comput|translat)/i,
  /\b(zero|0)\s*(ms|millisecond|latency)\b/i,
  /\bno\s+(delay|latency|wait)\b/i,
  /\bimmediate(ly)?\s+(process|analyz|translat|comput)/i,
];

const SLA_VIOLATION_PATTERNS = [
  /\bguarantee[ds]?\s+(100%\s+)?(uptime|availability)/i,
  /\b(99\.999+|100)%\s+(uptime|availability)/i,
  /\bnever\s+(goes?\s+down|fail|timeout)/i,
  /\bzero\s+(downtime|failures)/i,
];

export function apiResponseTime(options: ApiResponseTimeOptions): Guard {
  return {
    name: 'api-response-time',
    version: '0.1.0',
    description: 'Detects unrealistic response time claims and SLA violations in text',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      for (const p of UNREALISTIC_PATTERNS) {
        if (p.test(text)) issues.push('unrealistic_claim');
      }
      for (const p of SLA_VIOLATION_PATTERNS) {
        if (p.test(text)) issues.push('sla_violation');
      }

      const triggered = issues.length > 0;
      return {
        guardName: 'api-response-time',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(issues.length / 3, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
